source/extensions/filters/http/router/config.cc

```diff
@@ -20,6 +22,8 @@ Http::FilterFactoryCb RouterFilterConfig::createFilterFactoryFromProtoTyped(
       proto_config));
 
   return [filter_config](Http::FilterChainFactoryCallbacks& callbacks) -> void {
+    callbacks.addStreamDecoderFilter(
```

Comment:
I don't know if we want to add the filter at the end of the filter-chain. IMHO, Router filter is a core filter which I think we should leave it alone as much as possible.  On the other hand, if there is no route for the vhost, some other filter  ahead of the router-filter may have already taken the send_local_reply action. Instead make this filter a standalone filter may make it more flexible, WDYT?

Reply:
I think I can do what you are asking, will look into this.

Reply:
@stevenzzzz: I made OnDemand a first-class filter. I opted for not having users configure it (WDYT?) and instead make it always the first one to be called: https://github.com/envoyproxy/envoy/pull/8617/files#diff-c68454405a6c92916edd63682d33246fR346